### PR TITLE
Internal joint type as string

### PR DESCRIPTION
**Pull request by tetov**
### What type of change is this?

Tweak to handling of internal attribute in class.

### Checklist

- ~~I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).~~
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- ~~I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.~~
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tetov** [comment]
I don't see a reason to store the type as an integer. Not sure if Joint.FIXED etc makes sense if the types are strings, but would like some input on that.

**Comment by tetov** [comment]
## Reason for PR

Tried to create a joint like this
```
joint = Joint("name", Joint.REVOLUTE, "parent_joint", "child_joint")
```
which doesn't work since the class is written to accept a string as type.

**Review by gonzalocasas — CHANGES_REQUESTED** [review]
This is a breaking change.  So it would most definitely need a line in the changelog.
However, I don't think it's justified enough to go for a breaking change only to resolve a relatively minor syntactic inconvenience. A simpler and far less disruptive change would be to allow both `int` and `str` in the constructor of `Joint` and then check if it's one or the other.

**Comment by beverlylytle** [comment]
I worry that changing the `Joint` class constants from ints to strings will break things in `compas_fab`.  Also, the `Joint` constants are mimicked in `Configuration`.

**Comment by tetov** [comment]
Ok, I'll go with your suggestion @gonzalocasas. Thanks for the input!

**Comment by tetov** [comment]
`invoke test` from compas_fab main runs without errors with this branch installed.

**Comment by beverlylytle** [comment]
And `invoke test --doctest --codeblock`?

**Comment by tetov** [comment]
> 
> 
> And `invoke test --doctest --codeblock`?

Just finished with errors. But I just realized that I should run docker compose, haha!

EDIT: I'll report back when the containers are up and the tests have ran.

File: tests/compas/robots/test_joint.py [hunk 1]
```diff
@@ -9,6 +11,28 @@
...
 
 
+def test_constructor_joint_type_str():
+    for type_ in Joint.SUPPORTED_TYPES:
+        assert Joint("joint", type_, "parent_joint", "child_joint")
+
+
+def test_constructor_joint_type_int():
+    for type_ in range(len(Joint.SUPPORTED_TYPES)):
+        assert Joint("joint", type_, "parent_joint", "child_joint")
```
Comment (gonzalocasas) [inline tests/compas/robots/test_joint.py]: I would make these tests more verbose:
```suggestion
    assert Joint("joint", 'revolute', "parent_joint", "child_joint")
    assert Joint("joint", 'continuous', "parent_joint", "child_joint")
    assert Joint("joint", 'prismatic', "parent_joint", "child_joint")
    assert Joint("joint", 'fixed', "parent_joint", "child_joint")


def test_constructor_joint_type_int():
    assert Joint("joint", Joint.REVOLUTE, "parent_joint", "child_joint")
    assert Joint("joint", Joint.CONTINUOUS, "parent_joint", "child_joint")
    assert Joint("joint", Joint.PRISMATIC, "parent_joint", "child_joint")
    assert Joint("joint", Joint.FIXED, "parent_joint", "child_joint")
```

File: tests/compas/robots/test_joint.py [hunk 2]
```diff
@@ -9,6 +11,28 @@
...
+
+
+def test_constructor_joint_type_guard_str():
+    with pytest.raises(ValueError):
+        Joint("joint", "synovial", "femur", "tibia")
+
+
+def test_constructor_joint_type_guard_int():
+    joint_types = [-1, len(Joint.SUPPORTED_TYPES)]  # just outside acceptable range
```
Comment (gonzalocasas) [inline tests/compas/robots/test_joint.py]: Again, keep it simple, just do a single `Joint('joint', 150, 'femur', 'tibia')`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -15,6 +15,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 
 * Extended `compas.topology.astar_shortest_path` to work on `compas.datastructures.Mesh` and `compas.datastructures.Network`.
+* Joint type in the `compas.robots.Joint` constructor can now be be defined as
+  an index of `Joint.SUPPORTED_TYPES` as well as a member of
+  `Joint.SUPPORTED_TYPES`
```
Comment (gonzalocasas) [inline CHANGELOG.md]: More concise:
```suggestion
* Allow str or int as joint type in `compas.robots.Joint` constructor
```

**Review by gonzalocasas — APPROVED** [review]
lgtm
only a couple of suggestions

**Comment by tetov** [comment]
@beverlylytle 
> And `invoke test --doctest --codeblock`?

Passes!

```
(compas_dev) C:\Users\a\repos\compas_fab>inv test --doctest
============================= test session starts =============================
platform win32 -- Python 3.8.8, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: C:\Users\a\repos\compas_fab, configfile: setup.cfg, testpaths: src, tests
plugins: cov-2.12.0
collected 121 items

src\compas_fab\backends\pybullet\client.py .                             [  0%]
src\compas_fab\backends\pybullet\utils.py s                              [  1%]
src\compas_fab\backends\ros\client.py ...                                [  4%]
src\compas_fab\backends\ros\messages\geometry_msgs.py ..                 [  5%]
src\compas_fab\backends\vrep\client.py ....                              [  9%]
src\compas_fab\backends\vrep\planner.py .                                [  9%]
src\compas_fab\backends\vrep\backend_features\vrep_forward_kinematics.py . [ 10%]
                                                                         [ 10%]
src\compas_fab\robots\constraints.py .........                           [ 18%]
src\compas_fab\robots\inertia.py .                                       [ 19%]
src\compas_fab\robots\planning_scene.py ........                         [ 25%]
src\compas_fab\robots\rfl.py .                                           [ 26%]
src\compas_fab\robots\robot.py ...........................               [ 48%]
src\compas_fab\robots\tool.py ......                                     [ 53%]
src\compas_fab\robots\ur5.py .                                           [ 54%]
src\compas_fab\robots\wrench.py .......                                  [ 60%]
src\compas_fab\sensors\base.py s                                         [ 61%]
src\compas_fab\sensors\baumer.py ss                                      [ 62%]
tests\backends\test_tasks.py ..                                          [ 64%]
tests\backends\ros\test_local_cache_info.py ..                           [ 66%]
tests\backends\ros\messages\test_std_msgs.py ...                         [ 68%]
tests\robots\test_duration.py .......                                    [ 74%]
tests\robots\test_robot.py ......................                        [ 92%]
tests\robots\test_semantics.py .....                                     [ 96%]
tests\robots\test_trajectory.py ....                                     [100%]

=========================== short test summary info ===========================
SKIPPED [4] ..\..\scoop\apps\miniconda3\current\envs\compas_dev\lib\site-packages\_pytest\doctest.py:448: all tests skipped by +SKIP option
======================= 117 passed, 4 skipped in 12.57s =======================
argv[0]=
```

**Review by beverlylytle — APPROVED** [review]
LGTM!

**Comment by beverlylytle** [comment]
I'll resolve the conflicts and then merge.

**Pull request merged**